src/bootman/bootman.c

```diff
@@ -446,6 +457,208 @@ bool boot_manager_set_uname(BootManager *self, const char *uname)
         return self->have_sys_kernel;
 }
 
+static bool boot_manager_is_inird_valid_name(const char *name)
+{
+        if (strlen(name) < 3) {
+                return false;
+        }
+
+        return isdigit(name[0]) && isdigit(name[1]);
+}
+
+static int boot_manager_sort_initrd_name(const void *name1, const void *name2)
+{
+        return strcmp((char *)name1, (char *)name2);
+}
+
+bool boot_manager_get_initrd_freestanding(BootManager *self)
+{
+        autofree(DIR) *initrd_dir = NULL;
+        NcArray *dir_files = NULL;
+        struct dirent *ent = NULL;
+        struct stat st = { 0 };
+
+        if (!self || !self->initrd_freestanding_dir) {
+                return false;
+        }
+
+        dir_files = nc_array_new();
+        OOM_CHECK_RET(dir_files, false);
+
+        initrd_dir = opendir(self->initrd_freestanding_dir);
+        if (!initrd_dir) {
+                LOG_ERROR("Error opening %s: %s", self->initrd_freestanding_dir, strerror(errno));
+                nc_array_free(&dir_files, NULL);
+                return false;
+        }
+
+        while ((ent = readdir(initrd_dir)) != NULL) {
+                char *initrd_name = NULL;
+                autofree(char) *path = NULL;
+
+                path = string_printf("%s/%s", self->initrd_freestanding_dir, ent->d_name);
```

Comment:
So in this case you are allocating a new string each time this loops meaning that all the previous allocations leak.

Reply:
the pointer is being allocated in the nc_array, just when nc_array_add fails it is lost, then I added a free for this case

Reply:
The value of the path pointer is not being stored in the nc_array_add so will leak memory when multiple initrds are found.

Reply:
but it is autofree, it means that it will be free when it get out of context, right?

Reply:
The last value that is in path but not any value ever in path. So if you have path = 0x00 on loop one and then path = 0x08 on loop two, the memory in 0x08 will be freed but 0x00 will not.

Reply:
the end of context of the path variable is in the end of the loop, then free is called before the new loop starts, that is why that variable is declared inside the loop

Reply:
Oh you are so right, scope local, sorry X(.